src/metabase/util/secret.clj

```diff
@@ -0,0 +1,27 @@
+(ns metabase.util.secret
+  (:import (java.io Writer)))
+
+(set! *warn-on-reflection* true)
+
+;; Define a protocol for secrets to make things harder to accidentally expose.
+(defprotocol ISecret
+  (expose [this] "Expose the secret"))
+
+(defrecord Secret [value-fn]
+  ISecret
+  (expose [_this] (value-fn))
+  Object
+  (toString [_this] "<< REDACTED SECRET >>"))
```

Comment:
I wonder if this should be translated, in case it pops up on the UI somewhere? That's not going to happen right now, of course, but if this is a utility namespace it's difficult to forsee future use cases.

Probably an edge case though, so nbd either way.

Reply:
I think translation makes sense... but do we use `trs` or `tru`? It's actually kind of hard to say where these strings are more likely to end up, since the intent is that they won't end up anywhere... 🤔

I think if I had to guess I'd say it's slightly more likely for them to end up in server logs (or other telemetry data) than in the UI (it seems like it would take some intentional effort to put one in the UI, whereas you could just accidentally send a map as context on a log). So I'll use `trs` unless you think it makes sense to use `tru`.

Reply:
I'm merging this with `trs` for now, we can change to `tru` in a followup PR if desired.